File: linkedList/linkedListCycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

bool checkCur(struct ListNode* cur, struct ListNode* head){
    struct ListNode* check = head;
    while (check != cur){
        if (check == cur -> next){
            return true;
        }
        check = check -> next;
    }
    if (check == cur -> next){
        return true;
    }
    return false;
}

bool hasCycle(struct ListNode *head) {
    /*
    while we iterate through the linked list, for each node, we check if the cur node points to a previously explored node
    This code worked but was very slow:/
    */
    // base case of linked list of size 0 or uncy
    if (head == NULL || head -> next == NULL){
        return false;
    }

    struct ListNode* cur = head;
    while (cur != NULL){
        if (checkCur(cur, head)){
            return true;
        }
        cur = cur -> next;
    }
    return false;
}

File: linkedList/linkedListCycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fast = head.next if head else None
        while fast and fast != head:
            head = head.next
            fast = fast.next
            fast = fast.next if fast else fast
        return True if fast else False